File: projects/2026-011-sunken-village/07_publish/youtube/generate-thumbnails.py
#!/usr/bin/env python3
"""Overlay Korean thumbnail headlines onto 1280x720 backgrounds."""

from pathlib import Path

from PIL import Image, ImageDraw, ImageFont

HERE = Path(__file__).resolve().parent
WIDTH = 1280
HEIGHT = 720
FONT_PATH = "/System/Library/Fonts/AppleSDGothicNeo.ttc"

VARIANTS = [
    {
        "source": "thumbnail-a-background.png",
        "output": "thumbnail-a.png",
        "side": "left",
        "size": 80,
        "lines": [("철거된 종탑", "#ffffff"), ("종이 울렸다", "#ff4d5f")],
    },
    {
        "source": "thumbnail-b-background.png",
        "output": "thumbnail-b.png",
        "side": "right",
        "size": 76,
        "lines": [("42가구 이주", "#ffffff"), ("마을은 그대로", "#ff4d5f")],
    },
]


def load_font(size):
    for index in (7, 6, 0):
        try:
            return ImageFont.truetype(FONT_PATH, size, index=index)
        except OSError:
            continue
    raise FileNotFoundError(FONT_PATH)


def cover(path):
    image = Image.open(path).convert("RGB")
    scale = max(WIDTH / image.width, HEIGHT / image.height)
    resized = image.resize(
        (int(image.width * scale), int(image.height * scale)),
        Image.Resampling.LANCZOS,
    )
    left = (resized.width - WIDTH) // 2
    top = (resized.height - HEIGHT) // 2
    return resized.crop((left, top, left + WIDTH, top + HEIGHT))


def render(variant):
    canvas = cover(HERE / variant["source"])
    overlay = Image.new("RGBA", (WIDTH, HEIGHT), (0, 0, 0, 0))
    draw = ImageDraw.Draw(overlay)
    is_left = variant["side"] == "left"
    panel_x = 28 if is_left else 690
    for index in range(562):
        t = index / 561
        alpha = int(210 * (1 - 0.62 * t)) if is_left else int(210 * (0.28 + 0.62 * t))
        draw.line(
            [(panel_x + index, 145), (panel_x + index, 470)],
            fill=(0, 0, 0, alpha),
        )
    accent_x = 52 if is_left else 1198
    draw.rectangle([accent_x, 176, accent_x + 9, 414], fill=(227, 27, 54, 255))
    font = load_font(variant["size"])
    x = 82 if is_left else 1170
    for (text, color), y in zip(variant["lines"], (250, 370)):
        box = draw.textbbox((0, 0), text, font=font)
        text_w = box[2] - box[0]
        text_h = box[3] - box[1]
        tx = x if is_left else x - text_w
        ty = y - text_h // 2
        for dx in range(-6, 7):
            for dy in range(-6, 7):
                if dx * dx + dy * dy <= 36:
                    draw.text((tx + dx, ty + dy), text, font=font, fill=(0, 0, 0, 250))
        draw.text((tx, ty), text, font=font, fill=color)
    out = Image.alpha_composite(canvas.convert("RGBA"), overlay).convert("RGB")
    out.save(HERE / variant["output"], "PNG")
    print(f"Created {variant['output']}")


if __name__ == "__main__":
    for variant in VARIANTS:
        render(variant)

File: projects/2026-010-sixth-guest/07_publish/youtube/generate-thumbnails.mjs
import fs from "node:fs";
import path from "node:path";
import {fileURLToPath} from "node:url";
import {createCanvas, loadImage} from "canvas";

const here = path.dirname(fileURLToPath(import.meta.url));
const width = 1280;
const height = 720;

const variants = [
  {
    source: "thumbnail-a-background.png",
    output: "thumbnail-a.png",
    side: "left",
    lines: [
      {text: "5명 체크인", color: "#ffffff"},
      {text: "CCTV엔 6명", color: "#ff334f"},
    ],
  },
  {
    source: "thumbnail-b-background.png",
    output: "thumbnail-b.png",
    side: "right",
    lines: [
      {text: "다섯 명인데", color: "#ffffff"},
      {text: "컵은 6개", color: "#ff334f"},
    ],
  },
];

function drawCover(ctx, image) {
  const scale = Math.max(width / image.width, height / image.height);
  const sourceWidth = width / scale;
  const sourceHeight = height / scale;
  const sourceX = (image.width - sourceWidth) / 2;
  const sourceY = (image.height - sourceHeight) / 2;
  ctx.drawImage(
    image,
    sourceX,
    sourceY,
    sourceWidth,
    sourceHeight,
    0,
    0,
    width,
    height,
  );
}

function drawHeadline(ctx, variant) {
  const isLeft = variant.side === "left";
  const panelX = isLeft ? 28 : 690;
  const panelWidth = 562;
  const gradient = ctx.createLinearGradient(
    isLeft ? panelX : panelX + panelWidth,
    0,
    isLeft ? panelX + panelWidth : panelX,
    0,
  );
  gradient.addColorStop(0, "rgba(0, 0, 0, 0.78)");
  gradient.addColorStop(1, "rgba(0, 0, 0, 0.30)");
  ctx.fillStyle = gradient;
  ctx.fillRect(panelX, 145, panelWidth, 325);

  const accentX = isLeft ? 52 : 1198;
  ctx.fillStyle = "#e31b36";
  ctx.fillRect(accentX, 176, 9, 238);

  ctx.font = "800 82px 'Apple SD Gothic Neo', sans-serif";
  ctx.textAlign = isLeft ? "left" : "right";
  ctx.textBaseline = "middle";
  ctx.lineJoin = "round";
  ctx.lineWidth = 12;
  ctx.strokeStyle = "rgba(0, 0, 0, 0.96)";
  const x = isLeft ? 82 : 1170;
  const yPositions = [250, 370];
  variant.lines.forEach((line, index) => {
    ctx.strokeText(line.text, x, yPositions[index]);
    ctx.fillStyle = line.color;
    ctx.fillText(line.text, x, yPositions[index]);
  });
}

for (const variant of variants) {
  const image = await loadImage(path.join(here, variant.source));
  const canvas = createCanvas(width, height);
  const ctx = canvas.getContext("2d");
  drawCover(ctx, image);
  drawHeadline(ctx, variant);
  fs.writeFileSync(path.join(here, variant.output), canvas.toBuffer("image/png"));
  console.log(`Created ${variant.output}`);
}
